### PR TITLE
fix: replace use effect with use layout effect in initial value hook

### LINKED ISSUE #64: Hydration with `useFieldInitialValue` in Next.js renders twice - first empty, then filled input.
I believe that currently the `useFieldInitialValue` is not optimized for SSR, as it fails to render a filled input on the first paint.
I suspect that the `useEffect` might be the cause.
So far I've switched to `useHydrateAtoms` from `jotai/utils` which fixed the issue. https://github.com/form-atoms/field/commit/9264c90ae873cddca49aa660720d22b82a073173#diff-57ff5d92fdd79e6888455c22013fd0924199b51e87f1be3075f7c8b9cc848c47R5

The `hydrateAtoms` will set the atoms only once, while the `useFieldInitialValue` was recently updated to permit multiple updates. https://github.com/form-atoms/form-atoms/issues/52

The multiple updates I believe are also vital for Next.js. My use case is that after a form is saved, I call `router.refresh()`, where the router is the `useRouter()` from `nextjs/navigation`. This successfully re-initialized my form with the changes from #52, but it fails to work (expected) with the `useHydrateAtoms`.

Potential solutions
- use the `useHydrateAtoms` in the `useFieldInitialValue`
- update `useFieldInitialValue` to set the first value outside of the `useEffect`?

Beware that the `useHydrateAtoms` might not be straighforward - with my linked implementation for `useHydrateField` I'm experiencing HTML content mismatch, as multiple `listAtom` items result to multiple DOM nodes. So after the `router.refresh()` when item in a list was added/removed, the FE jotai store is stalled (since `useHydrateAtoms` is permited only once) and the BE renders different amount of items, so a mismatched HTML. Here I can clear the jotai store, but likely that won't help as the internal `hydrateMap` preventing the duplicate hydration is not resetable from outside...  https://github.com/pmndrs/jotai/blob/2d760920f5ebea05abaa8e072cee578fa125ca9b/src/react/utils/useHydrateAtoms.ts#L22C7-L22C18
So maybe even that could be made resetable in the jotai lib, but that would require manual user cache invalidation ,e.g. just after the `router.refresh()` call in my case.

### PR DESCRIPTION
**Pull request by jaredLunde**
closes #64

### CODE REVIEW

**Comment by MiroslavPetrik** [comment]
This scares me as it equals to useHydrateState with its options.dangerouslyHydrate flag as on.

I was experimenting today a bit
https://github.com/pmndrs/jotai/discussions/2337#discussioncomment-8146448

I would test it first. Isnt the store.set calling setState? Via useReducer i think. So that equals setting state in render which is wrong.

**Comment by jaredLunde** [comment]
It's actually perfectly ok to set state in render https://github.com/facebook/react/blob/9a5f28dbedc2b2edd01d216e347ab24b8efa73fc/packages/use-subscription/src/useSubscription.js#L42-L56

See also: https://swizec.com/blog/react-can-update-state-during-render/#:~:text=Calling%20the%20set%20function%20during,information%20from%20the%20previous%20renders.

**Comment by MiroslavPetrik** [comment]
So I've tested it locally for my listAtom submission, and it works great that it is able not only render on first render, but also to rehydrate after item is added/removed as indicated in the original issue. Good idea to use the `setState` in render.

But, the limitation "only allowed from within the currently rendering component" won't be a problem? 

I have a composition of 
```tsx
import {List} from "@fom-atoms/field";

const AddressListField = ({initialValue}) => {
  useFieldInitialValue(addresses, initialValue); // useHydrateField previously
  return (<List field={addresses} />)
}
```

which as I thought previously do causes a warning (same as useHydrateAtoms when used with the `dangerouslyForceHydrate` flag.
```
Warning: Cannot update a component (`List`) while rendering a different component (`AddressListField`). To locate the bad setState() call inside `AddressListField`, follow the stack trace as described in https://reactjs.org/link/setstate-in-render
```

So yeah, this means that we can't place the hook to the parent of the component where we read the field value. This is ok, but I'm not sure if is ok for other scenarios like when we read the value in some sibling component. I guess it should work.

File: src/index.tsx [hunk 1]
```diff
@@ -1044,15 +1044,17 @@ export function useFieldInitialValue<Value>(
...
-      return;
-    }
-    if (!store.get(field.dirty)) {
-      store.set(field.value, initialValue);
-    }
+  if (initialValue === undefined) {
+    return;
+  }
+
+  if (!store.get(field.dirty) && initialValue !== store.get(field.value)) {
```
Comment (MiroslavPetrik) [inline src/index.tsx:1051]: I've just realized that for my list the value is array, which is never equal to the initial value, as it's a derived atom.
https://github.com/form-atoms/field/blob/3183dfd4ffda80e9d225582952dfb2be795509b7/src/atoms/list-atom/listAtom.ts#L187

maybe callback like `isFieldInitialized(initialValue, field, store)` can used

so I can supply my own, and either compare array deeply, or keep the last initialValue in a separate atom on the list, or other strategy...

**Comment by jaredLunde** [comment]
fixed to handle arrays

File: src/index.tsx [hunk 2]
```diff
@@ -1039,20 +1039,36 @@ export function useFieldErrors<Value>(
...
+}
+
+function defaultValuesAreEqual(a: unknown, b: unknown): boolean {
+  if (Array.isArray(a) && Array.isArray(b)) {
+    return (
+      a.length === b.length && a.every((v, i) => defaultValuesAreEqual(v, b[i]))
+    );
+  }
+
+  return Object.is(a, b);
```
Comment (MiroslavPetrik) [inline src/index.tsx:1071]: we need to check first for object is, as with the initialValue, we will unnecessarily do deep array compare, while very often those will be the same array.

`object.is(a,b) || arraysAreEqual(a,b)`

**Comment by MiroslavPetrik** [comment]
When I hit reset in the codesandbox:
image

**Comment by jaredLunde** [comment]
Yeah. I see that now -unfortunate.

**Comment by jaredLunde** [comment]
Nah, no chance I'm doing that. Caveats and options are footguns. I'm inclined to leave it in useEffect with the addition of the array check

**Pull request merged**